p2p/peer.go

```diff
@@ -499,87 +539,38 @@ func PeerMetrics(metrics *Metrics) PeerOption {
 }
 
 func (p *peer) metricsReporter() {
-	for {
-		select {
-		case <-p.metricsTicker.C:
-			status := p.mconn.Status()
-			var sendQueueSize float64
-			queues := make(map[byte]int, len(status.Channels))
-			for _, chStatus := range status.Channels {
-				sendQueueSize += float64(chStatus.SendQueueSize)
-				queues[chStatus.ID] = chStatus.SendQueueSize
-			}
-
-			p.metrics.PeerPendingSendBytes.With("peer_id", string(p.ID())).Set(sendQueueSize)
-			schema.WritePendingBytes(p.traceClient, string(p.ID()), queues)
-		case <-p.Quit():
-			return
-		}
-	}
 }
 
-//------------------------------------------------------------------
-// helper funcs
-
-func createMConnection(
-	conn net.Conn,
-	p *peer,
-	reactorsByCh map[byte]Reactor,
-	msgTypeByChID map[byte]proto.Message,
-	chDescs []*cmtconn.ChannelDescriptor,
-	onPeerError func(Peer, interface{}),
-	config cmtconn.MConnConfig,
-) *cmtconn.MConnection {
-
-	onReceive := func(chID byte, msgBytes []byte) {
-		reactor := reactorsByCh[chID]
-		if reactor == nil {
-			// Note that its ok to panic here as it's caught in the conn._recover,
-			// which does onPeerError.
-			panic(fmt.Sprintf("Unknown channel %X", chID))
+func (p *peer) StartReceiving() error {
+	for {
+		stream, err := p.conn.AcceptStream(context.Background())
+		if err != nil {
+			p.Logger.Error("failed to accept stream", "err", err.Error())
+			return err
 		}
-		mt := msgTypeByChID[chID]
-		msg := proto.Clone(mt)
-		err := proto.Unmarshal(msgBytes, msg)
+		var chID byte
+		err = binary.Read(stream, binary.BigEndian, &chID)
 		if err != nil {
-			panic(fmt.Errorf("unmarshaling message: %s into type: %s", err, reflect.TypeOf(mt)))
+			p.Logger.Error("failed to read channel ID", "err", err.Error())
+			return err
 		}
-
-		if w, ok := msg.(Unwrapper); ok {
-			msg, err = w.Unwrap()
-			if err != nil {
-				panic(fmt.Errorf("unwrapping message: %s", err))
+		// start accepting data
```

Comment:
Before continuing to listen for new messages, it must check whether the given `chID` is supported.